pkg/kubelet/cm/cpumanager/cpu_manager.go

```diff
@@ -359,7 +367,7 @@ func (m *manager) removeStaleState() {
 	assignments := m.state.GetCPUAssignments()
 	for podUID := range assignments {
 		for containerName := range assignments[podUID] {
-			if _, ok := activeContainers[podUID][containerName]; !ok {
+			if _, ok := terminatingContainers[podUID][containerName]; ok {
```

Comment:
I'm not expert in pod lifecycle, so bear with me: the intention of the old code (IIUC) is to clean up resources previously assigned by non-active pod. This includes (IIUC again) terminating pods, but also pod terminated long ago. Do we free resources allocated to terminated pods in the new flow?

Reply:
yes, it should, so under terminating, pods will fall all pods to with the pod worker set the `startedTerminating` or if the pod already does not exist.

It is the logic to get terminating pods:
https://github.com/kubernetes/kubernetes/blob/d6f2473d081fba803898a6240b841c2d7f2bd1f3/pkg/kubelet/pod_workers.go#L419

Reply:
Is this really true? This function will return a list of *all* pods that are either currently terminating or have terminated at any point in the past? Keeping (and looping through) that list of *all* terminated pods since the beginning of time seems incorrect.

Reply:
mm it has the same problem as activePods, the pod worker still does not have any information regarding the pod :(

Reply:
Can you explain the race we are attempting to avoid with this?

It sounds like with the refactor it's now possible that a CPU assignment could exist in the `assignments` variable above, but the pod/container that it has been assigned to won't have made its way into the list of "active" pods returned by the `activePods()` function. So we end up removing the CPU assignment as "stale" if the reconcile loop happens to run inside this small window where the assignment has been made but the pod is not yet seen as "active".

Reply:
@klueska this is the most concise summary of the issue @cynepco3hahue is working on: https://github.com/kubernetes/kubernetes/issues/103952#issuecomment-887859214

Reply:
Got it. What @smarterclayton says makes sense. I think we can track state from admission -> start of each container and check against that + active pods to decide if we should remove the state as stale of not.

Reply:
Agreed, we are converging towards that direction.

Reply:
So what allowed this to work before? Were pods still in admission part of the active pods list previously?

Reply:
Yep, it was the buggy behavior that we used under our managers.